src/darwin/Framework/CHIP/CHIPDevice.h

```diff
@@ -49,6 +72,91 @@ NS_ASSUME_NONNULL_BEGIN
+                        attributeId:(NSUInteger)attributeId
+                        clientQueue:(dispatch_queue_t)clientQueue
+                         completion:(CHIPDeviceResponseHandler)completion;
+
+/**
+ * Write to attribute in a designated attribute path
+ *
+ * @param completion  response handler will receive either value or error. value will be an NSArray object with NSDictionary
+ * elements. Each NSDictionary will have "endpointId", "clusterId", "attributeId" and "status" keys. "endpointId", "clusterId",
+ * "attributeId" and "status" will be mapped to NSNumber objects. "status" with 0 value indicates success and non-zero value
+ * indicates failure.
+ */
+- (void)writeAttributeWithEndpointId:(NSUInteger)endpointId
+                           clusterId:(NSUInteger)clusterId
+                         attributeId:(NSUInteger)attributeId
+                               value:(id)value
+                         clientQueue:(dispatch_queue_t)clientQueue
+                          completion:(CHIPDeviceResponseHandler)completion;
+
+/**
+ * Invoke a command with a designated command path
+ *
+ * @param commandFields   command fields object. The object must be an NSDictionary object representing attribute value
+ *                      as described in the readAttributeWithEndpointId:clusterId:attributeId:clientQueue:responseHandler: method.
+ *                      The attribute must be a Structure, i.e., the NSDictionary "type" key must have the value "Structure".
+ *
+ * @param completion  response handler will receive either value or error. value will be an NSArray object with NSDictionary
+ * elements. Each NSDictionary will have "endpointId", "clusterId", "commandId", "status" and "responseData" keys. "endpointId",
+ * "clusterId", "attributeId" and "status" will be mapped to NSNumber objects. "status" with 0 value indicates success and non-zero
+ * value indicates failure. "responseData" key will be included only when "status" key has 0 value and there is response data for
+ * the command. "responseData" key value will be an NSDictionary object representing attribute value as described in the
+ * readAttributeWithEndpointId:clusterId:attributeId:clientQueue:responseHandler: method.
+ */
+- (void)invokeCommandWithEndpointId:(NSUInteger)endpointId
+                          clusterId:(NSUInteger)clusterId
+                          commandId:(NSUInteger)commandId
+                      commandFields:(id)commandFields
+                        clientQueue:(dispatch_queue_t)clientQueue
+                         completion:(CHIPDeviceResponseHandler)completion;
+
+/**
+ * Subscribe an attribute in a designated attribute path
+ *
+ * @param reportHandler   handler for the reports. Note that only the report handler by the last call to this method per the same
+ * attribute path will receive reports. Report handler will receive either value or error. value will be an NSDictionary object. The
```

Comment:
This isn't a valid note. I used an old report registry which kept only the last registered report handler but switched to ReadClient later because the former would break things between multiple apps.